File: Heap/Java/The Number of the Smallest Unoccupied Chair.java
// Problem link: https://leetcode.com/problems/the-number-of-the-smallest-unoccupied-chair/description
// Time: O(n log n)
// Space: O(n)
class Solution {
    public int smallestChair(int[][] times, int targetFriend) {
        int size = times.length, chair;
        Integer[] indexes = new Integer[size];

        for(int idx = 0; idx < size; idx++) {
            indexes[idx] = idx;
        }

        Arrays.sort(indexes, (a, b) -> Integer.compare(times[a][0], times[b][0]));
        PriorityQueue<Integer> minChair = new PriorityQueue<>();
        PriorityQueue<int[]> minLeaving = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));

        for(int chr = 0; chr < size; chr++) {
            minChair.add(chr);
        }

        for(int index : indexes) {
            while(!minLeaving.isEmpty() && minLeaving.peek()[0] <= times[index][0]) {
                minChair.add(minLeaving.poll()[1]);
            }

            chair = minChair.poll();

            if(index == targetFriend) {
                return chair;
            }

            minLeaving.add(new int[] {times[index][1], chair});
        }

        return -1;
    }
}

File: Heap/Python/The Number of the Smallest Unoccupied Chair.py
# Problem link: https://leetcode.com/problems/the-number-of-the-smallest-unoccupied-chair/description
# Time: O(n log n)
# Space: O(n)
class Solution:
    def smallestChair(self, times: List[List[int]], targetFriend: int) -> int:
        indexes = sorted([idx for idx in range(len(times))], key = lambda x : times[x][0])
        minHeapOne, minHeapTwo = [time for time in range(len(times))], []

        for idx in indexes:
            while minHeapTwo and minHeapTwo[0][0] <= times[idx][0]:
                heapq.heappush(minHeapOne, heapq.heappop(minHeapTwo)[1])
            
            chair = heapq.heappop(minHeapOne)

            if idx == targetFriend:
                return chair
            
            heapq.heappush(minHeapTwo, (times[idx][1], chair))

        return -1
